### PR TITLE
Use alloca in MVM_string_memmem_grapheme32str...

### PR DESCRIPTION
**Pull request by MasterDuke17**
based on how it's done in knuth_morris_pratt_string_index().

I tested this with a one-liner that just does some `.contains()` and
`.starts-with()` over a spesh log that's ~200k lines long. Timing it
(with `MVM_SPESH_BLOCKING=1`) reported ~2.36s before this change and
~2.26 after. Additionally, heaptrack reported ~4.7M calls to allocation
functions and ~1.8M temporary allocations before, and ~3.0M calls to
allocation functions and ~110k temporary allocations after.

NQP builds ok and passes `make m-test` and Rakudo builds ok and passes `make m-test m-spectest`.

### CODE REVIEW

File: src/strings/ops.c [hunk 1]
```diff
@@ -544,9 +544,17 @@ static MVMint64 MVM_string_memmem_grapheme32 (MVMThreadContext *tc, MVMGrapheme3
 static MVMint64 MVM_string_memmem_grapheme32str (MVMThreadContext *tc, MVMString *Haystack, MVMString *needle, MVMint64 H_start, MVMStringIndex H_graphs, MVMStringIndex n_graphs) {
     MVMGrapheme32 *needle_buf = NULL;
     MVMint64 rtrn;
+    int needle_is_malloced = 0;
     if (needle->body.storage_type != MVM_STRING_GRAPHEME_32) {
         MVMStringIndex i;
-        needle_buf = MVM_malloc(needle->body.num_graphs * sizeof(MVMGrapheme32));
+        size_t needle_size = (1 + n_graphs) * sizeof(MVMGrapheme32);
+        /* Allocate max 8K onto the stack, otherwise malloc */
+        if (needle_size < 3000)
```
Comment (jnthn) [inline src/strings/ops.c:552]: Comment says 8K, allocation amount is 3K?

Comment (MasterDuke17) [reply src/strings/ops.c:552]: That was just copied from `knuth_morris_pratt_string_index()`, looks like the comment is wrong there too. I'll correct the comments and force push.

**Comment by MasterDuke17** [comment]
Callgrind reports ~21,050,000,000 instructions before and ~20,408,000,000 instructions after.

File: src/strings/ops.c [hunk 2]
```diff
@@ -544,24 +544,32 @@ static MVMint64 MVM_string_memmem_grapheme32 (MVMThreadContext *tc, MVMGrapheme3
...
         else {
-            for (i = 0; i < needle->body.num_graphs; i++) {
+            for (i = 0; i < n_graphs; i++) {
                 needle_buf[i] = needle->body.storage.blob_8[i];
             }
         }
     }
     rtrn = MVM_string_memmem_grapheme32(tc, Haystack->body.storage.blob_32, needle_buf ? needle_buf : needle->body.storage.blob_32, H_start, H_graphs, n_graphs);
-    if (needle_buf) MVM_free(needle_buf);
+    if (needle_buf && needle_is_malloced) MVM_free(needle_buf);
```
Comment (zhuomingliang) [inline src/strings/ops.c:572]: Change to `needle_is_malloced && needle_buf ` will be better. For most case, if `needle_is_malloced ` is true, it should not  test `needle_buf` again.

Comment (MasterDuke17) [reply src/strings/ops.c:572]: Actually, it doesn't even need the `&& needle_buf`, right? If `needle_is_malloced` is set, we called MVM_malloc with a size >= 3000. If that were to fail it'd blow up earlier. Ah, just looked and `knuth_morris_pratt_string_index` just does the `*_is_malloced` test, so switching all to just that.

**Pull request merged**